### PR TITLE
Replace Timeout methods with branching "run" methods

### PR DESCRIPTION
**Pull request by katcharov**
JAVA-5376

This removes all production usages of the methods. Remaining work pending feedback regarding approach. Please look for `TODO-CSOT` when reviewing, and treat these as PR comments - some are notes for work remaining to be done, some indicate places where the approach runs into issues, and some are questions/confirmations regarding CSOT.

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/time/Timeout.java
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/time/Timeout.java:75]: Suggest keeping this but renaming it so its usage is clear that its a point in time check. eg: `standaloneCheckHasExpired()` (also advocate adding `standaloneCheckHasExpired(TimeUnit)`). 

I think its easier than having to use the `run` for this usecase.

**Review by rozza — COMMENTED** [review]
This its looking good and its going to be a good improvement.

Added some comments and feedback.

File: driver-core/src/main/com/mongodb/internal/CheckedConsumer.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/CheckedConsumer.java:23]: If code inside a lambda throws a checked exception, the usual functional interfaces will complain, but these ones will "throws" it out of the `checkedRun` method, which follows the pattern of [checkedWithLock](https://github.com/jyemin/mongo-java-driver/blob/3a21597e5822fac003c0962f8df062b7816eec8a/driver-core/src/main/com/mongodb/internal/Locks.java#L39) (for which CheckedSupplier was originally added).

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:129]: I think changing to MS won't fully prevent a race, because for example if we check hasExpired at exactly 1ms (not expired) but a few nanoseconds elapse from then until the point we get remaining, remaining(ms) will round down returning 0. But I think the race is fixed, because since the `remaining` methods have been removed from prod, all the places where we could have gotten a 0 now explicitly check for expiry again and specify distinct expired/infinite branches based on only that single check.

Will comment on hasExpired below.

File: driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:532]: Worth noting that this messy usage (and at least one other) would be avoided under the async API. It would be a throw, which would cease execution, and the exception would be handled by the  API.

File: driver-core/src/test/functional/com/mongodb/internal/connection/DefaultConnectionPoolTest.java
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/internal/connection/DefaultConnectionPoolTest.java:536]: Note that the `hasExpired` method appears to be used exclusively for stopping execution. The only places where we don't stop execution by throwing are [here](https://github.com/mongodb/mongo-java-driver/pull/1349/files#diff-e80f234d8c415657233df477aad007b21149b6c258e633449ff3f776c0ba8b69R273), and for determining `isLastAttempt` (via the inline under `TimeoutContext#hasExpired`). If we can find a (prod) usage where a throw wouldn't improve things, I see no harm in keeping, but otherwise I think we should steer into forcing a throw.

An aside: I don't understand how `lastAttempt` works. We can't know if an attempt is the last attempt until the end of an attempt. If we check during the attempt (to add error labels), we could get "not last", not add the labels, but at the end of the attempt the timer has run out and actually it "is last", so we end... but without the requisite labels. We should check this - maybe the RetryState cleanup should be done sooner, to avoid possible bugs.

Regarding the TimeUnit, I don't think that we should fail fast (preemptively expire), since some ensuing operations (internal code, externals that take nanos) might complete within nanoseconds, and I see no reasonable way to enforce or consistently reason about what might come after.

File: driver-sync/src/main/com/mongodb/client/internal/TimeoutHelper.java
Comment (katcharov) [reply driver-sync/src/main/com/mongodb/client/internal/TimeoutHelper.java:40]: Current docs for MongoCollection#withTimeout do not mention MS being the lowest granularity. Is this where we wanted to fail an assertion if the user provided (0 < value < 1ms)? I can't recall what technical reason there was for needing to be at or above 1ms...

File: driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:524]: Refactored to to use existsAndExpired to mirror `receiveCommandMessageResponse`. This uses nanoseconds, but the two locations should be aligned. (See other comments for discussion regarding: I believe no race created; and not "failing fast".)

File: driver-core/src/main/com/mongodb/internal/connection/LoadBalancedCluster.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/connection/LoadBalancedCluster.java:228]: Added an `ifExpired` method. Perhaps for greater consistency it should be onExpired, or this one ifExistsAndHasExpired?

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/KeyManagementService.java
Comment (katcharov) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/KeyManagementService.java:166]: Updated the TODO, will leave it in so that it is caught in the final PR.

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/gridfs/GridFSUploadPublisherImpl.java
Comment (katcharov) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/gridfs/GridFSUploadPublisherImpl.java:268]: Switched to nanos, on the principle that we should use lowest granularity unless an external method forces us to truncate to milliseconds.

File: driver-reactive-streams/src/test/unit/com/mongodb/reactivestreams/client/internal/gridfs/GridFSUploadPublisherImplTest.java
Comment (katcharov) [reply driver-reactive-streams/src/test/unit/com/mongodb/reactivestreams/client/internal/gridfs/GridFSUploadPublisherImplTest.java:97]: Addressed in other comment. Removed this TODO.

File: driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java:269]: Updated to throw into the ensuing catch.

File: driver-core/src/main/com/mongodb/internal/time/Timeout.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/time/Timeout.java:75]: Replied to similar topic [here](https://github.com/mongodb/mongo-java-driver/pull/1349#discussion_r1534736064)

File: driver-core/src/main/com/mongodb/internal/CheckedConsumer.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/CheckedConsumer.java:23]: Ah good to know - thanks

File: driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java
Comment (katcharov) [reply driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java:135]: Perhaps the underlying problem that you identify, of the first and last params, is solved by adding `nullAsInfinite`, which I have pushed up:

```
    return Timeout.nullAsInfinite(operationTimeout).checkedRun(NANOSECONDS,
                    () -> socket.getInputStream(),
                    (ns) -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)),
                    () -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)));
    }
```

This makes it explicit why the first and last are the same, and removes redundant code. It also facilitates extracting that wrapping, which I think belongs [outside of the method](https://github.com/mongodb/mongo-java-driver/pull/1215/files#r1537863409). It also means we can get rid of static variants. 

I found it helpful when implementing to think of params as going from "least expired to most expired" (infinite never expires, then remaining, then expired). I am not sure how helpful this is for others, but I do not recall needing to check param order when making changes. Perhaps this "order mnemonic" addresses remaining readability issues? (Early on, @stIncMale and I briefly discussed the fluent approach. I think it would take more work and design, and would have its own downsides.)

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:159]: The check above:
```
Long timeoutMS = timeoutSettings.getTimeoutMS();
if (timeoutMS == null) { ...
```
Is it correct that this is equivalent to hasTimeout, which is equivalent to checking if timeout != null? (Or, can timeout be non-null, but hasTimeout returns false, due to some sort of hierarchical override?)

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:184]: Looks like this is used by at least `TimeoutContext#getConnectTimeoutMs`. I refactored to return a timeout via `Timeout calculateConnectTimeoutMs`, which then forces the PR's typical pattern of using the branching run to invoke external methods.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:211]: This is tricky refactor, but I think we end up with?:
```
Timeout t = timeout != null 
        ? timeout 
        : Timeout.expiresIn(timeoutSettings.getMaxTimeMS(), MILLISECONDS);
t = t.shortenBy(minRoundTripTimeMS, MILLISECONDS);
return t.run(MILLISECONDS,
        () -> 0L,
        (ms) -> ms,
        () -> {
            throw createMongoRoundTripTimeoutException();
        });
```
That is, we start with the legacy `maxTimeMS` if we don't have a timeout, but whichever one we use is shortened by the round trip time (I think your snippet above does not shorten legacy time). Please confirm.

I refactored to push the run into a helper "put" method. This seemed to be the only use of maxTimeMS, which left `getMaxTimeMS` with only test usages. I can clean that up if you'd like, otherwise can be part of previously used JAVA-5385.

File: driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java:135]: I agree with @vbabanin using named handlers over unnamed parameters makes the code self explaining / readable. That said this pattern is similar to how a reactive stream {{Subscriber}} has named events, but then libraries (Reactor) often add helpers eg: `subscribe(Subscriber)` gets extended out to `subscribe(OnNext), subscribe(OnNext, onError), subscribe(OnNext, OnError, OnComplete)` .  Once you start using it and get familiar with it you often use the convience methods. 🤷

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:198]: If there is no timeout then you defer to `setting.getMaxTimeMS`. That figure is used as is and doesn't take round trip into account.

File: driver-sync/src/main/com/mongodb/client/internal/TimeoutHelper.java
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/TimeoutHelper.java:40]: Cool, will ensure the docs are clarified. There will be confirmation that it is at least 1ms.  Technical reasoning other than the spec, is the server methods only take MS values for their timeouts.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:135]: This still needs to be done in MS. If there is less than 1MS left then it should not continue, eg shouldnt send anything on the wire or should wrap any seen timeout errors as CSOT errors.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:181]: The following seems clearer to me:

```java
Timeout.min(
        Timeout.expiresInWithZeroAsInfinite(alternativeTimeoutMS, MILLISECONDS),
        Timeout.nullAsInfinite(timeout));
```

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/time/TimePoint.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/time/TimePoint.java:106]: I refactored the implementing method, but kept the base method. The intent is to remove the base method, once we have consensus on the approach, and then this can be changed back.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:28]: Please see https://github.com/mongodb/mongo-java-driver/pull/1215/files#r1543521955 (@rozza or @vbabanin)

File: driver-core/src/test/functional/com/mongodb/internal/connection/DefaultConnectionPoolTest.java
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/internal/connection/DefaultConnectionPoolTest.java:536]: The above aside does seem to be a bug, which I have recorded as [a comment on the final PR](https://github.com/mongodb/mongo-java-driver/pull/1215/files#r1543542029).

File: driver-core/src/main/com/mongodb/internal/time/Timeout.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/time/Timeout.java:254]: Removed.

Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/time/Timeout.java:254]: I have not added the null assertions that we discussed. CheckedSupplier is used in Locks; if you search for `if (localPendingReader != null) {`, you will see a case where it is being used to return a null value.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java]: Updated

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:164]: This change assumes that it is impossible for timeout to be null when `timeoutSettings.getTimeoutMS()` returns a non-null value. Please confirm.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:163]: This timeoutOrAlternative method should ideally be inlined, with `getMaxCommitTimeMS`, `getReadTimeoutMS`, and `getWriteTimeoutMS` encapsulating the result using lambdas, as was done in `runMaxTimeMSTimeout`. This requires more associated refactoring than seems appropriate for this PR.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:52]: Resetting to null to designate a default seemed clearer, and made the default call more obvious in the context of the logic where this was used (which is one place).

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:185]: Considered and tried various alternatives:

1. Returning the remaining time leaks internals (which has led to bugs).
2. Returning a live Timeout causes unified tests to fail, because a timeout of 100ms will be sent to the server as 99ms as soon as any nanoseconds have elapsed.
3. Could return a fixed value under a new type of timeout, which would pass that fixed value to its run method (that is, it would not be a timeout, but a configuration value provider); but this is a stretch, can't be compared for "earlier" against other timeouts. 

This chosen approach follows the pattern of existing timeout runs, and without leaking internals, breaking tests, or abusing the Timeout type.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/TimeoutContext.java:164]: It appears that in practice, based on usages, timeout is never null in such cases. The constructors are always invoked where the timeout is derived via `startTimeout`, or have a non-null value. Also, if this were not the case, then I think `resetTimeout` has a bug, because it derives the timeout based on `timeoutSettings.getTimeoutMS`.

Refactored a bit.

File: driver-core/src/main/com/mongodb/internal/time/TimePoint.java
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/time/TimePoint.java]: The `nanos == null` comparison is still used in `TimePoint` instead of the `isInfinite` method.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: config/spotbugs/exclude.xml
Comment (katcharov) [reply config/spotbugs/exclude.xml:282]: Refactoring moved the issue into a lambda, and IIRC there is no way to match on a lambda due to tool limitations. This "match" applies to all methods in the class.

File: driver-core/src/main/com/mongodb/internal/time/Timeout.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/time/Timeout.java:97]: That is just the way to create a timeout. The extra "OrInfiniteIfNegative" wording is meant to remind of the semantics (which seems more important than possible confusion, as has happened here), rather than suggest that there is an alternate "timeoutAfter" method that should be used. I have added a comment.

**Review by stIncMale — APPROVED** [review]
I have not reviewed everything (mostly the new API). Hopefully, the combination of all reviews covers the changes in full.

File: config/spotbugs/exclude.xml
Comment (rozza) [reply config/spotbugs/exclude.xml:282]: Ah, good to know

**Comment by katcharov** [comment]
Merged CSOT, 1 conflict: removed `timeoutRemainingMS` and its temp fix (see 4e97f21, "added temp fix for timeoutRemainingMS (scheduled to be removed)").

File: driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java
Comment (vbabanin) [reply driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java:134]: I understand the rationale behind the "least expired to most expired" parameter organization. However, concerns arise regarding its long-term clarity and potential for error as this implicit understanding might not be as apparent in the future.

The approach with fluent API, while initially demanding more effort, promises greater maintainability in the future, making it easier to understand, modify, and debug.

Nevertheless, I respect the current direction chosen by the team and am willing to align with the consensus.

Seeing that @stIncMale and @ross are comfortable with the current implementation and considering the existing approvals on the PR, I believe you can go ahead and merge the PR without needing my explicit approval.

Comment (ross) [reply driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java:134]: > @ross

Probably not the Ross you're looking for.

File: driver-core/src/main/com/mongodb/internal/TimeoutContext.java
Comment (vbabanin) [inline driver-core/src/main/com/mongodb/internal/TimeoutContext.java:214]: [Optional] Currently, this method accepts a pre-calculated maxTimeMS. One of the original intents behind considering a supplier was to enable operations to calculate maxTimeMS right before the command is sent, potentially reducing timing errors. In my local testing indicates that this delayed calculation method can indeed be more precise, reducing the error margin to about 1-3 milliseconds compared to pre-calculated values.

For now, there's no need for a method with delayed maxTimeMS calculation, but this could change. I suggest adding the Javadoc to reflect that it accepts pre-calculated values and to mention the future possibility of using a supplier lambda for more precise timing, ensuring we remember this option moving forward.

**Review by vbabanin — APPROVED** [review]

**Pull request merged**
